docs/application.rst

```diff
@@ -1,80 +1,75 @@
-Application Structure
-=====================
+The Warehouse codebase
+======================
 
-Note: this is a brain dump and its contents may be moved to a more
-appropriate location eventually.
+Warehouse uses the
+`Pyramid <https://docs.pylonsproject.org/projects/pyramid/en/latest/index.html>`__
+web framework, the
+`SQLAlchemy <https://docs.sqlalchemy.org/en/latest/>`__ ORM, and
+`Postgres <https://www.postgresql.org/docs/>`__ for its database.
+Warehouse's front end uses Jinja2 templates.
 
-At the moment it *just* lists the legacy structure and none of the intended
-new structure.
+The application exists within two `Docker <https://docs.docker.com/>`__
+containers, one of which contains static files for the website, and the
+other which contains the Python web application code running in a
+virtual environment and the database. In the development environment,
+`Docker Compose <https://docs.docker.com/compose/overview/>`__ manages
+running the containers and the connections between them.
 
-The following documents the current URLs in the legacy PyPI application.
+Since Warehouse was built on top of a pre-existing database and
+developers had to fit our ORM to the existing tables, some of the code
+in the ORM may not look like code from SQLAlchemy’s documentation. There
+are some places where joins are done using name-based logic instead of a
+foreign key (but this may change in the future).
 
-============= =================================================================
-URL           Purpose
-------------- -----------------------------------------------------------------
-/             Redirect to /pypi
-/pypi         Legacy PyPI application. See below.
-/daytime      Legacy mirroring support
-/security     Page giving contact and other information regarding site security
-/id           OpenID endpoint
-/oauth        OAuth endpoint
-/simple       Simple API as given in :doc:`api-reference/legacy`
-/packages     Serve up a package file
-/mirrors      Page listing legacy mirrors (not to be retained)
-/serversig    Legacy mirroring support (no-one uses it: not to be retained)
-/raw-packages nginx implementation specific hackery (entirely internal; not to
-              be retained)
-/stats        Web stats. Whatever. Probably dead.
-/local-stats  Package download stats. All the legacy mirrors have this.
-/static       Static files (CSS, images) in support of the web interface.
-============= =================================================================
+Warehouse also uses `Pyramid's hybrid URL traversal and
+dispatch <https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hybrid.html>`__.
+Using factory classes, URLs are pre-populated before the view is
+requested.
```

Comment:
I would rephrase this line a bit, I'm not sure 'URLs are pre-populated' makes sense, maybe something like "Using factory classes, resources are provided directly to the views based on the URL pattern"